### ISSUE TITLE
When pushing a single json file to a chang edit via gerrit_api.put, library errors

### ISSUE THREAD

**Issue by gi11es** [dpursehouse/pygerrit2#54]
I think the response content type sniffing doesn't work if you push a json file. Adding any other file to a change is fine, but when I try to push a json file that ends in .json to a change, I get:

`gerrit_api.put('/changes/%s/edit/blog.json' % changeid, data=contents)`

> ERROR:root:Invalid json content: 
> Traceback (most recent call last):
>   File "checkupdate.py", line 58, in <module>
>     publish_update(output.splitlines())
>   File "checkupdate.py", line 44, in publish_update
>     add_file_to_change_edit(changeid, filepath, contents)
>   File "checkupdate.py", line 23, in add_file_to_change_edit
>     gerrit_api.put(api_url, data=contents)
>   File "/usr/local/lib/python2.7/dist-packages/pygerrit2/rest/__init__.py", line 179, in put
>     decoded_response = _decode_response(response)
>   File "/usr/local/lib/python2.7/dist-packages/pygerrit2/rest/__init__.py", line 60, in _decode_response
>     return json.loads(content)
>   File "/usr/lib/python2.7/json/__init__.py", line 339, in loads
>     return _default_decoder.decode(s)
>   File "/usr/lib/python2.7/json/decoder.py", line 364, in decode
>     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
>   File "/usr/lib/python2.7/json/decoder.py", line 382, in raw_decode
>     raise ValueError("No JSON object could be decoded")
> ValueError: No JSON object could be decoded

However, the call actually succeeded from Gerrit's perspective, and the file was added to the change edit.

**Comment by dpursehouse** [comment]
Sorry for the delayed response.

I've written a test for this use case (see 3c29786e674bc352157cade84e68101bbe23cb0c) but it is passing.

Can you have a look at that and see if it accurately reproduces your use case?  My test only includes a trivial json content; maybe it needs to be more elaborate?

**Comment by gi11es** [comment]
But this unit tests isn't getting a response from an actual gerrit server, is it? The bug lies in how the library parses the actual response, not in sending the request to the gerrit server.

**Comment by dpursehouse** [comment]
The test is making the request (and getting the response) against a real
gerrit server that is running in docker. It tests against the latest
versions of gerrit 2.14, 2.15, and 2.16.

**Comment by gi11es** [comment]
Could be version-specific, maybe? I see that we're running Gerrit 2.15.8-7-g780f1f2b91 and that the live tests pull other versions.

The tests looks definitely just like the code I have that triggers the ValueError.

I wrote this small repro script against our own Gerrit server:

```
from pygerrit2 import GerritRestAPI, HTTPBasicAuthFromNetrc, GerritReview

url = 'https://gerrit.wikimedia.org/r/'

auth = HTTPBasicAuthFromNetrc(url=url)
gerrit_api = GerritRestAPI(url=url, auth=auth)

changeinput = {'project': 'performance/docroot', 'subject': 'Testing pygerrit2 bug', 'branch': 'master', 'topic': 'foo'}
result = gerrit_api.post('/changes/', data=changeinput)
changeid = result['id']

contents = """{"foo" : "bar"}"""

api_url = '/changes/%s/edit/file.json' % changeid
gerrit_api.put(api_url, data=contents)
```

Which yields:

```
gilles@perfbot:~/pygerrit2$ python repro.py 
[19-02-25 10:51:57]-[pygerrit2-ERROR] Invalid json content: 
Traceback (most recent call last):
  File "repro.py", line 15, in <module>
    gerrit_api.put(api_url, data=contents)
  File "/home/gilles/pygerrit2/pygerrit2/rest/__init__.py", line 186, in put
    decoded_response = _decode_response(response)
  File "/home/gilles/pygerrit2/pygerrit2/rest/__init__.py", line 67, in _decode_response
    return json.loads(content)
  File "/usr/lib/python2.7/json/__init__.py", line 339, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python2.7/json/decoder.py", line 364, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python2.7/json/decoder.py", line 382, in raw_decode
    raise ValueError("No JSON object could be decoded")
ValueError: No JSON object could be decoded
```

I tried adding 2.15.8 to the pytest.fixture params of your test, but "make livetests" seems to get stuck on pulling the image. I have a terrible internet connection today, so it might very well be a problem on my end.

**Comment by gi11es** [comment]
If you want to try the exact version we're running, here's the commit of our mirror/minor fork currently deployed on that host: https://github.com/wikimedia/operations-software-gerrit/commit/780f1f2b9196a95d96253f1aa93d62aa4b5811cc

**Comment by gi11es** [comment]
I was finally able to run the tests by pointing your code to 2.15.8 and they pass...

```
livetests.py::TestGerritAgainstLiveServer::test_put_with_json_dict[2.15.8] 
Pulling image gerritcodereview/gerrit:2.15.8
⠋
Container started:  4d7c9cf789
Waiting to be ready...
PASSED
livetests.py::TestGerritAgainstLiveServer::test_put_with_data_dict[2.15.8] PASSED
livetests.py::TestGerritAgainstLiveServer::test_post_with_data_dict_and_no_data[2.15.8] PASSED
livetests.py::TestGerritAgainstLiveServer::test_post_with_json_dict[2.15.8] PASSED
livetests.py::TestGerritAgainstLiveServer::test_put_with_data_as_string[2.15.8] PASSED
livetests.py::TestGerritAgainstLiveServer::test_put_json_content[2.15.8] PASSED
livetests.py::TestGerritAgainstLiveServer::test_get_base64_data[2.15.8] PASSED
livetests.py::TestGerritAgainstLiveServer::test_get_patch_zip[2.15.8] PASSED
livetests.py::TestGerritAgainstLiveServer::test_put_with_no_content[2.15.8] PASSED
livetests.py::TestGerritAgainstLiveServer::test_review[2.15.8] PASSED

======================================================================================================== 10 passed in 31.41 second
```

It might have to do with the specific version we're running, or the plugins we run, maybe?

**Comment by gi11es** [comment]
I've figured out what's tripping up the library, but I don't know where it's coming from.

Our server returns a 204 response on change PUTs, whose content-type headers matches the contents of the file that was pushed.

As a result, the library attempts to parse the response body, which is empty since it's a 204.

I don't know what's causing our Gerrit server to have that content-type when apparently according to your tests, vanilla Gerrit doesn't do that, but a plugin we have installed might be responsible.

**Comment by dpursehouse** [comment]
Thanks for the additional info.  It could well be caused by a plugin; I'll try to look into this some more during the week.

**Comment by dpursehouse** [comment]
I've fetched wikimedia/operations-software-gerrit@780f1f2 locally and built that, but I'm not able  to reproduce the issue against it.  I wonder if it's not enough to have those plugins loaded, but also need to configure gerrit in a specific way?

**Comment by dpursehouse** [comment]
Although I couldn't reproduce the issue, I've made some changes which should at least prevent it from trying to convert an empty response to json.

Could you please try against your server with the tip of the issue-54 branch?

https://github.com/dpursehouse/pygerrit2/tree/issue-54

If you enable debug logging you should see logs of the response content type and status code.

**Comment by gi11es** [comment]
I've verified that the issue is fixed when at the tip of that branch. Thank you!

**Comment by dpursehouse** [comment]
Thanks for confirming.  I'll make a new release of the library after updating the tests to run against the latest Gerrit versions.  Should be within the next week or so.

**Comment by dpursehouse** [comment]
PS, I would be interested to hear about it if you figure out what was causing this on the Gerrit side.

One thing that occurred to me: is your instance of Gerrit running behind a proxy or in a container that might be modifying the response headers?

**Comment by gi11es** [comment]
Behind Apache, I believe, which indeed might do something funny with the response headers.

**Comment by dpursehouse** [comment]
I've just released version 2.0.9 which includes the fix for this issue.
